### DISCUSSION TITLE
Livewire V3 still exposes model IDs

### DISCUSSION THREAD

**Discussion by francoism90** [livewire/livewire#6334] [3. Feature ideas]
I really love Livewire, however one of the issues is exposing the model-id and class when using model binding.
An alternative is to not use this, however you will lose typehints, and cannot use `wire:model`.

I created a previous PR for V2 to encrypt this: https://github.com/livewire/livewire/pull/5742

However in V3 the logic has been changed. Luckily, it only exposes the model class morph, however this still can be too much information.

The solution would be to encrypt the full meta response, which seems to be possible, however this may lead to performance lost.

**Comment by DanielSpravtsev** [comment]
As I understood, Livewire v3 has gone from using model bindings. Now it is preferable to use Form Objects for that purpose. 

Maybe @joshhanley can correct me if it's wrong

**Reply by francoism90** [reply]
@DanielSpravtsev 

This applies to forms, but not to route model binding (
https://livewire.laravel.com/docs/components#using-route-model-binding):

```php
public function PostController extends Component
{
    public Post $post;
}
```

Is this also discourage?

**Comment by PhiloNL** [comment]
If you don't want to expose it without making database changes, I would recommend looking into [Hashids](https://github.com/vinkla/hashids). For example:

```php
class EditPost extends Component
{
    #[Locked]
    public $postId;

    #[Rule('required|min:3')] 
    public $title;

    #[Rule('required|min:3')] 
    public $body;

    public function mount($postId)
    {
        $this->postId = $postId;
        $this->title = $this->post()->title;
        $this->body = $this->post()->body;    
    }

    public function save()
    {
        $validated = $this->validate();

        $this->post()->update($validated);
    }

    #[Computed]
    public function post()
    {
        return Post::findByHashidOrFail($this->postId);
    }
}
```

**Reply by francoism90** [reply]
@PhiloNL Thanks, but I did this already in Livewire V2.

The problem is that you will lose typings, and with this, it's maybe better to use the new form builder.

If I'm not mistaken, the computed value can still be read by JS, as Livewire still uses the model connection?

It would be great if the whole meta key could be encrypted, as this would hide the class name, model ID and possible other details. I'm using this in V2 and so far I didn't found any issues.

**Reply by francoism90** [reply]
Also, I'm already use unique IDs, that's why I'm forcing the model route key. This may also be useful, when some uses the slug as route key.

But encryption would prevent the need of using the route key, and simple obscure the details.

**Reply by tpetry** [reply]
The idea they are proposing is quite interesting. With a new attribute you could encrypt the values when being sent as state.

```php
class EditPost extends Component
{
    #[Encrypted] // implies #[Locked]
    public Post $post;
}

```

**Reply by calebporzio** [reply]
`#[Encrypted]`is pretty interesting. 

Not sure it would imply locked though. In this case models are locked anyways, but I'm picturing `#[Encrypted]` being applicable to any public property, even mutable ones and it would just encrypt the metadata.

Because it has to encypt and decrypt before Livewire does it's thing it won't be as simple as normal attributes but that doesn't have to stop us. I think this could be a good way to put the power in the users hands

**Reply by ProgrammerZ** [reply]
Surprised that Livewire did not do this automatically on public models. `#[Encrypted]` would make life much easier.

**Comment by francoism90** [comment]
It seems Filament handles this differently:
https://github.com/filamentphp/filament/blob/f040ba92b76af3042067f677eaf87265e1eff813/packages/panels/src/Resources/Pages/Concerns/InteractsWithRecord.php

They also use the route logic to resolve the model:
https://github.com/filamentphp/filament/blob/3.x/packages/panels/src/Resources/Resource.php#L154

**Comment by francoism90** [comment]
If anyone is interested, I have changed the model binding with a package: https://github.com/foxws/wireuse

The problem with encryption, is that it has performance implications for simple route binding. However, it would be really helpful for more sensitive data.

**Reply by ProgrammerZ** [reply]
If speed is of concern, you can use something like Sqids or Optimus. They should not be used for very sensitive data, but it's still better than just plain model ids on the page. Real encryption by default with ability use other drivers.

**Reply by francoism90** [reply]
@ProgrammerZ I use [prefixed-ids](https://github.com/spatie/laravel-prefixed-ids) myself. The package uses the route-key, so you could indeed use Sqids/Optimus/slug if that's a requirement. :)

**Comment by moisish** [comment]
I have been thinking about this and I believe Livewire should use the routeKeyName rather than the key of the model. This way you can expose the uuid,slug etc which you have on the url. If its good for your urls, its good for your livewire stuff

If @calebporzio agrees I can work out a PR

**Reply by francoism90** [reply]
I'm already doing this. It should be compatible with most things, but still it could be a breaking change. Some packages bind using the key, not the route key.
